src/toolkits/activity_classification/activity_classifier.cpp

```diff
@@ -464,6 +464,130 @@ gl_sframe activity_classifier::predict_per_window(gl_sframe data,
   return result;
 }
 
+gl_sframe activity_classifier::predict_topk(gl_sframe data,
+                                            std::string output_type, size_t k,
+                                            std::string output_frequency) {
+  // check valid input arguments
+  if (output_type != "probability" && output_type != "rank") {
+    log_and_throw(output_type +
+                  " is not a valid option for output_type.  "
+                  "Expected one of: probability, rank");
+  }
+  if (output_frequency != "per_row" && output_frequency != "per_window") {
+    log_and_throw(output_frequency +
+                  " is not a valid option for output_frequency.  "
+                  "Expected one of: per_row, per_window");
+  }
+
+  // data inference
+  std::unique_ptr<data_iterator> data_it =
+      create_iterator(data, /* requires_labels */ false, /* is_train */ false,
+                      /* use_data_augmentation */ false);
+  gl_sframe raw_preds_per_window = perform_inference(data_it.get());
+
+  // argsort probability to get the index (rank) for top k class
+  // if k is greater than the class number, set it to be class number
+  flex_list class_labels = read_state<flex_list>("classes");
+  k = std::min(k, class_labels.size());
+  auto argsort_prob = [=](const flexible_type& ft) {
+    const flex_vec& prob_vec = ft.get<flex_vec>();
+    std::vector<size_t> index_vec(prob_vec.size());
+    std::iota(index_vec.begin(), index_vec.end(), 0);
+    auto compare = [&](size_t i, size_t j) {
+      return prob_vec[i] > prob_vec[j];
+    };
+    std::nth_element(index_vec.begin(), index_vec.begin() + k, index_vec.end(),
```

Comment:
I just realized that you need to check k is not larger than prob_vec.size(), which means the k should be larger than the number of class labels in the current data iterator. Otherwise, the behavior will be undefined. Welcome to Cpp world. You should put a pr to prevent potential segfault.

The logic should be like, 
```
if (!raw_preds_per_window["preds"].empty())
k = std::min(k, raw_preds_per_window["preds"][0].size());
```
Let's have DASSERT to make sure each cell of raw_preds_per_window["preds"] has same size, if the sframe is not empty.

Reply:
I think I do check k's range at 491 ?